dotnet/src/Connectors/Connectors.Memory.Pinecone/PineconeVectorStoreRecordMapper.cs

```diff
@@ -65,27 +65,25 @@ internal sealed class PineconeVectorStoreRecordMapper<TRecord> : IVectorStoreRec
     /// <summary>
     /// Initializes a new instance of the <see cref="PineconeVectorStoreRecordMapper{TDataModel}"/> class.
     /// </summary>
-    /// <param name="keyProperty">A property info object that points at the key property for the current model, allowing easy reading and writing of this property.</param>
-    /// <param name="dataProperties">A list of property info objects that point at the data properties in the current model, and allows easy reading and writing of these properties.</param>
-    /// <param name="vectorProperties">A list of property info objects that point at the vector properties in the current model, and allows easy reading and writing of these properties.</param>
-    /// <param name="storagePropertyNames">A dictionary that maps from a property name to the configured name that should be used when storing it.</param>
-    public PineconeVectorStoreRecordMapper(PropertyInfo keyProperty, List<PropertyInfo> dataProperties, List<PropertyInfo> vectorProperties, Dictionary<string, string> storagePropertyNames)
+    /// <param name="vectorStoreRecordDefinition">The record definition that defines the schema of the record type.</param>
+    public PineconeVectorStoreRecordMapper(
+        VectorStoreRecordDefinition vectorStoreRecordDefinition)
     {
-        Verify.True(vectorProperties.Count == 1, "There should be exactly one vector property in the data model.");
-
+        // Validate property types.
+        (PropertyInfo keyProperty, List<PropertyInfo> dataProperties, List<PropertyInfo> vectorProperties) = VectorStoreRecordPropertyReader.FindProperties(typeof(TRecord), vectorStoreRecordDefinition, supportsMultipleVectors: false);
         VectorStoreRecordPropertyReader.VerifyPropertyTypes([keyProperty], s_supportedKeyTypes, "Key");
-        VectorStoreRecordPropertyReader.VerifyPropertyTypes(dataProperties, s_supportedDataTypes, "Data", s_supportedEnumerableDataElementTypes);
+        VectorStoreRecordPropertyReader.VerifyPropertyTypes(dataProperties, s_supportedDataTypes, s_supportedEnumerableDataElementTypes, "Data");
         VectorStoreRecordPropertyReader.VerifyPropertyTypes(vectorProperties, s_supportedVectorTypes, "Vector");
 
+        // Assign.
         this._keyPropertyInfo = keyProperty;
         this._dataPropertiesInfo = dataProperties;
         this._vectorPropertyInfo = vectorProperties[0];
-        this._storagePropertyNames = storagePropertyNames;
 
-        foreach (var property in dataProperties.Concat(vectorProperties).Concat([keyProperty]))
-        {
-            this._jsonPropertyNames[property.Name] = VectorStoreRecordPropertyReader.GetJsonPropertyName(JsonSerializerOptions.Default, property);
-        }
+        // Get storage names and store for later use.
+        var properties = VectorStoreRecordPropertyReader.SplitDefinitionAndVerify(typeof(TRecord).Name, vectorStoreRecordDefinition, supportsMultipleVectors: false, requiresAtLeastOneVector: true);
```

Comment:
In line 73 we use `VectorStoreRecordPropertyReader.FindProperties` which returns `keyProperty`, `dataProperties` and `vectorProperties` either from C# type or record definition object and in following lines we are doing a validation with `VerifyPropertyTypes` method. Here in line 84 we get these properties again, but only from record definition, and at this point record definition is already populated by user or from C# type. The question is - do we need to collect and verify these properties twice?

Also, in Qdrant logic I can see that there is `SplitDefinitionAndVerify` method and then `VerifyPropertyTypes` method calls, while here it's vice versa - we call `VerifyPropertyTypes` and then `SplitDefinitionAndVerify`. It would be nice to align these calls across all connectors, so connector developers will be able to follow the same pattern easily in the future.

Reply:
Note that the properties being returned are actually different in these two cases.  One provides PropertyInfo types from reflection and the other returns our VectorStoreRecordProperty object.  That said, I think there is a case for improving things a little bit further.  I have made some changes to that affect.  Please take a look.

Note that there will always be slight differences for each provider, because some require json property names, some require storage names, and some both.

In terms of the duplicated validation.  Splitting validation out of these methods are tricky.  I don't think it's too bad running the validation twice in some limited cases, since the alternative would probably be uglier, where we have to do a bunch of validation inline.  E.g. if we want to validate the number of keys found, we need to return a list of them, but then we need to validate and go to a single key property.  Easier to just do some validation in the helper method and return a single key to make the code here simpler.